v2/bigtable-cdc-to-hbase/src/main/java/com/google/cloud/teleport/v2/templates/transforms/ConvertChangeStream.java

```diff
@@ -0,0 +1,287 @@
+     */
+    private byte[] getLastCellColFamily(RowMutations hbaseMutations) throws IOException {
+      // het last mutation to get last column family from
+      Mutation lastMutation =
+          hbaseMutations.getMutations().get(hbaseMutations.getMutations().size() - 1);
+      CellScanner scanner = lastMutation.cellScanner();
+      // get to last cell
+      Cell cell = scanner.current();
+      while (scanner.advance()) {
+        cell = scanner.current();
+      }
+      return CellUtil.cloneFamily(cell);
+    }
+
+    /**
+     * Converts Bigtable ChangeStreamMutation to HBase RowMutations object.
+     *
+     * @param mutation changeStreamMutation
+     * @return Hbase RowMutations object
+     */
+    public static RowMutations convertToRowMutations(ChangeStreamMutation mutation)
+        throws Exception {
+      // Check for empty change stream mutation, should never happen.
+      if (mutation.getEntries().size() == 0) {
+        throw new Exception("Change stream entries list is empty.");
+      }
+
+      byte[] hbaseRowKey = mutation.getRowKey().toByteArray();
+      RowMutations hbaseMutations = new RowMutations(hbaseRowKey);
+      // Check and convert entries to hbase mutations
+      for (Entry entry : mutation.getEntries()) {
+        if (entry instanceof SetCell) {
+          hbaseMutations.add(convertSetCell(hbaseRowKey, entry));
+        } else if (entry instanceof DeleteCells) {
+          hbaseMutations.add(convertDeleteCells(hbaseRowKey, entry));
+        } else if (entry instanceof DeleteFamily) {
+          hbaseMutations.add(convertDeleteFamily(hbaseRowKey, entry));
+        } else {
+          // All change stream entry types should be supported.
+          throw new Exception("Change stream entry is not a supported type for conversion.");
+        }
+      }
+      return hbaseMutations;
+    }
+
+    private static Put convertSetCell(byte[] hbaseRowKey, Entry entry) {
+      SetCell setCell = (SetCell) entry;
+      // Convert timestamp to milliseconds
+      long ts = convertMicroToMilliseconds(setCell.getTimestamp());
+      Put put = new Put(hbaseRowKey, ts);
+      put.addColumn(
+          convertUtf8String(setCell.getFamilyName()),
+          setCell.getQualifier().toByteArray(),
+          setCell.getValue().toByteArray());
+
+      return put;
+    }
+
+    private static Delete convertDeleteCells(byte[] hbaseRowKey, Entry entry) {
+      DeleteCells deleteCells = (DeleteCells) entry;
+
+      // Convert timestamp to milliseconds
+      long ts = convertMicroToMilliseconds(deleteCells.getTimestampRange().getEnd());
+      // If RowMutation is created with an unbounded timestamp, the mutation is meant to delete
+      // all versions of a cell up to the point of operation.
+      // This behavior is approximated by casting the Hbase delete as deleting up to current point
+      // in time.
+      if (deleteCells.getTimestampRange() == TimestampRange.unbounded()) {
+        ts = Time.now();
```

Comment:
You should use the timestamp of CDCMutation. Basically time at which CBT accepted this delete. If you process this message after 3 days, you are deleting 3 extra days worth of mutations. This will cause divergence.

Reply:
Changed to use commit timestamp